uldbutil/uldbutil.1
.TH ULDBUTIL 1 "2012-08-21"
.SH NAME
uldbutil \- Utility for Accessing the Upstream LDM Database
.SH SYNOPSIS
.HP
.B
uldbutil
.HP
.B
uldbutil -d
.SH DESCRIPTION
.PP
This program access the database of upstream LDM processes. An upstream LDM
process is a child process of the local LDM server that's responsible for
feeding a remote downstream LDM in response to a request for data from the
downsteam LDM.
.PP
In the first form, this utility prints the contents of the upstream LDM
database to the standard
output stream. The format of the output is multiple lines each containing the
following fields:
.RS
.IR "pid protoVers type host fromTime toTime " { prodSpec "[,...]} mode"
.RE
where:
.RS
.TP
.I
pid
Process-identifier of the upstream LDM.
.TP
.I
protoVers
Protocol version (e.g., 5 or 6) of the upstream LDM.
.TP
.I
type
The type of the upstream LDM: "feeder" or "notifier".
.TP
.I
host
The Internet name of the computer on which the downstream LDM process is
running. This will normally be the fully-qualified hostname of the computer
unless the IP address can't be resolved into a hostname, in which case the IP
address will be printed.
.TP
.I fromTime
Earliest creation-time of matching data-products or \fBTS_BEGT\fP for the
beginning-of-time.
.TP
.I toTime
Latest creation-time of matching data-products or \fBTS_ENDT\fP for the
end-of-time.
.TP
.I prodSpec
Specification of the products requested by the downstream LDM in the form
.RS
.RS
.RI { feedSpec , " pattern" }
.RE
where:
.RS
.TP
.I feedSpec
The union of feedtypes of matching data-products.
.TP
.I pattern
Extended regular-expression for the product-identifier of matching
data-products.
.RE
.PP
There will normally be only one product-specfication.
.RE
.TP
.I mode
Transfer-mode of the connection to the downstream LDM. Either "primary" or
"alternate".
.RE
.PP
In the second form, this utility attempts to delete the upstream LDM 
database.
.SH EXIT STATUS
.PP
Error messages should be printed to the standard error stream if the utility
fails for any reason.
.TP
.B 0
Success.
.TP
.B 1
Incorrect usage (\fIi.e.\fP, command-line syntax error).
.TP
.B 2
The upstream LDM database doesn't exist.  This is likely because the LDM
server isn't running.
.TP
.B 3
The upstream LDM database exists but couldn't be accessed.
.SH "SEE ALSO"
.BR uldb (3)
.SH SUPPORT
.LP
If you have problems with this program, then you should first examine the 
LDM email archive for similar problems and how they were solved.
The email archive is available via the following URL:
.sp
.RS
\fBhttp://www.unidata.ucar.edu/software/ldm/\fP
.RE
.sp
If this does not suffice, then send an inquiry via email -- together will 
all relevant information -- to
.sp
.RS
\fBsupport-ldm@unidata.ucar.edu\fP
.RE


CHANGE_LOG
6.11.3.2 	2012-12-27 13:19:00-07:00
    uldb(3):
        Changed rejection by ULDB.

        Replaced ULDB rejection by ULDB reduction

        The upstream LDM database no longer rejects a subscription by a
        downstream LDM based on existing subscriptions from the same downstream
        host but, rather, reduces the subscription by the existing
        subscriptions (which might result in an empty subscrption -- which is,
        effectively, the same thing).

        Now, the ULDB only modifies the subscription by a downstream LDM if the
        transfer mode is primary. This is done because the ULDB feature
        interacts with the autoswitching feature. An autoswitch from alternate
        mode to primary mode risks receiving many products that won't make it
        into the product-queue because they've already been received due to the
        delay in termination of the (previous) upstream LDM in alternate mode.

    pqact(1):
        Added printing of product-ID to debug message

    ulog/log.c:
        Eliminated setting of _XOPEN_SOURCE to 500.  Necessary in order to
        build on SunOS 5.11 system using /opt/solstudio12.2/bin/c99.

    Installation:
        Adapted extractDecls(1) to SunOS 5.11's /usr/bin/awk.

        Fix creation of ~/data if it already exists and isn't a link

    Documentation:
        Added "pipe_put(): write error: ..." error-message

        Documented "-p pattern" option in pqsend(1).

        Documented use of uldbutil(1) for monitoring upstream LDM-s

        Updated to 2012 workshop:
            Made break-times indefinite in workshop web pages.

            Corrected class times

            Changed domain-name from guest.ucar.edu to fl-guest.ucar.edu

            Added configuration variable YEAR to workshop-specifics

        Replaced 6.9.8 with @VERSION@

    Misc:
        Makefile:
            Corrected "git status -a" to "git status -u"

6.11.2 	2012-12-17 11:11:20-07:00
    HOTFIX!

    LDM server:
        Differentiated between feeders and notifiers by the upstream LDM
        database (ULDB) when rejecting a request by a downstram LDM.

        Also removed all uldb_remove() calls except in cleanup() routine.

    ULDB module:
        Corrected use of RWL_EXIST in initLock() in misc/semRWLock.c.

    Makefile:
        Removed unnecessary "releaseCheck" target.

        Corrected "git status -a" to "git status -u".

6.11.1 	2012-10-05 10:21:05-06:00
    Makefile:
        Improved rules for releasing a new version:
            Removed targets "releaseCheck" and "ensureRelease".
            
            Corrected how "release" rule sets version in AC_INIT().

            Made "release" mandatory to prevent bad CHANGE_LOG from being
            released.

            Corrected "commit-check" rule: added "-Fq" to grep(1).

    Upstream LDM Database (uldbutil(1) & uldb(3)):
        Added pathname to uldb(3) initialization so that "make check" in
        the "protocol/" subdirectory doesn't interfere with a running LDM
        system.

        Improved logging by both uldb(3) and uldbutil(1).

    pqing(1):
        Corrected setting of logging file-descriptor (thanks Harry).

        Moved parsing of "-s" option to outside #ifdef conditional.

        Corrected version printed.

    Documentation:
        Corrected main index in tutorial.

        Added 5th channel example to NOAAPORT documentation.

        Corrected CHANGE_LOG for release 6.11.0.

        Added RELEASE_NOTES.

    Misc:
        Defined _XOPEN_SOURCE as 600 and corrected #includes of "config.h".

        Removed Eclipse-identified lint.

        Regenerated protocol/ldm_xdr.c.

        Improved some internal documentation.

        Removed unnecessary directories "pnglib/CVS" and "zlib/CVS".

6.11.0 	2012-09-24 16:13 -06:00
    Instructions:
        Improved preinstallation instructions: add modification of Red Hat
        net.ipv4.conf.default.rp_filter kernel parameter to preinstallation
        instructions.

        Corrected LDM configuration-file example: merged multi-line entry into
        single line.

    Installation:
        Improved configure(1) error-message: if the configuration-file for the
        system logging daemon isn't found, then the error-message will now
        indicate that the problem might be because it's not readable.

    NOAAPORT:
        Updated GRIB tables.

        Added noaaport/mainpage.h to repository.

        Fixed installation of NOAAPORT webpages

    ldmadmin(1):
        ldmadmin(1) calculates the number of slots from the size of the
        surf-queue. Modified to allow use of suffixes on size specification.

    ldmd(1):
        Added transitory upstream LDM database.

    uldbutil(1):
        Created uldbutil(1) utility to list local upstream LDM processes.

    pqact(1):
        Improved command-line parsing and usage message: two operands caused
        default configuration-file to be used, which is a latent bug.
        Error-messages during command-line parsing are printed to the
        standard-error stream, which doesn't exist in the LDM context.

        Merged Raytheon's changes from ldm-6.8.1

    pqing(1):
        Made the size of the largest expected data-product a command-line
        option (-s size).

    pqinsert(1):
        Added "-i" option to usage-message and improved message.

        Improved exit status.

    Logging:
        Modified log_start(): it no longer clears accumulated log messages.

        Moved filename and line number to start of log messages.

        Corrected addition of "localx.none" to system logging
        configuration-file when an entry comprises multiple lines.

        Demoted "Couldn't flush connection" log message from ERROR to NOTE.

    General:
        Removed some lint reported by Eclipse's C code analysis.

        Merged AWIPS-2 modifications from Raytheon

6.10.1 	2011-10-24 13:11:04-06:00
    Installation:

uldbutil/uldbutil.c
/**
 * Copyright 2012 University Corporation for Atmospheric Research.
 * All rights reserved.
 * See file ../COPYRIGHT for copying and redistribution conditions.
 *
 * This file contains the uldbutil(1) utility for accessing the upstream LDM
 * database.
 *
 * Created on: Aug 20, 2012
 * Author: Steven R. Emmerson
 */

#include <config.h>

#include <libgen.h>
#include <netinet/in.h>
#include <stdio.h>
#include <unistd.h>

#include "ldm.h"
#include "log.h"
#include "uldb.h"
#include "inetutil.h"
#include "ldmprint.h"
#include "prod_class.h"

static void printUsage(
        const char* progname)
{
    log_add("Usages:");
    log_add("  Print Database:     %s", progname);
    log_add("  Delete Database:    %s -d", progname);
    log_log(LOG_ERR);
}

/**
 * @retval 0    Success
 * @retval 1    Invocation error
 * @retval 2    The upstream LDM database doesn't exist
 * @retval 3    The upstream LDM database exists but couldn't be accessed
 */
int main(
        int argc,
        char* argv[])
{
    const char* const progname = ubasename(argv[0]);
    int status = 0;
    int delete = 0;

    (void) openulog(progname, LOG_NOTIME | LOG_IDENT, LOG_LDM, "-");
    (void) setulogmask(LOG_UPTO(LOG_NOTICE));

    {
        int ch;

        opterr = 0; /* Suppress getopt(3) error messages */

        while (0 == status && (ch = getopt(argc, argv, "d")) != -1) {
            switch (ch) {
            case 'd':
                delete = 1;
                break;
            default:
                LOG_START1("Unknown option: %c", optopt);
                printUsage(progname);
                status = 1;
                break;
            }
        }

        if (0 == status && optind < argc) {
            LOG_START0("Too many arguments");
            printUsage(progname);
            status = 1;
        }
    }

    if (0 == status) {
        if (delete) {
            if (status = uldb_delete(NULL)) {
                if (ULDB_EXIST == status) {
                    LOG_ADD0("The upstream LDM database doesn't exist");
                    log_log(LOG_INFO);
                    status = 2;
                }
                else {
                    LOG_ADD0("Couldn't open the upstream LDM database");
                    log_log(LOG_ERR);
                    status = 3;
                }
            }
        }
        else if (status = uldb_open(NULL)) {
            if (ULDB_EXIST == status) {
                LOG_ADD0("The upstream LDM database doesn't exist");
                LOG_ADD0("Is the LDM running?");
                log_log(LOG_NOTICE);
                status = 2;
            }
            else {
                LOG_ADD0("Couldn't open the upstream LDM database");
                log_log(LOG_ERR);
                status = 3;
            }
        }
        else {
            uldb_Iter* iter;

            if (status = uldb_getIterator(&iter)) {
                LOG_ADD0("Couldn't get database iterator");
                log_log(LOG_ERR);
                status = 3;
            }
            else {
                const uldb_Entry* entry;

                status = 0;

                for (entry = uldb_iter_firstEntry(iter); NULL != entry; entry =
                        uldb_iter_nextEntry(iter)) {
                    prod_class* prodClass;

                    if (uldb_entry_getProdClass(entry, &prodClass)) {
                        LOG_ADD0(
                                "Couldn't get product-class of database entry");
                        log_log(LOG_ERR);
                        status = 3;
                        break;
                    }
                    else {
                        const struct sockaddr_in* sockAddr =
                                uldb_entry_getSockAddr(entry);
                        char buf[2048];
                        const char* const type =
                                uldb_entry_isNotifier(entry) ?
                                        "notifier" : "feeder";

                        (void) s_prod_class(buf, sizeof(buf), prodClass);
                        (void) printf("%ld %d %s %s %s %s\n",
                                (long) uldb_entry_getPid(entry),
                                uldb_entry_getProtocolVersion(entry), type,
                                hostbyaddr(sockAddr), buf,
                                uldb_entry_isPrimary(entry) ? "primary" : "alternate");
                        free_prod_class(prodClass);
                    } /* "prodClass" allocated */
                } /* entry loop */

                uldb_iter_free(iter);
            } /* got database iterator */

            (void) uldb_close();
        } /* database opened */
    } /* correct invocation syntax */

    exit(status);
}
